source/agora/network/RPC.d

```diff
@@ -0,0 +1,498 @@
+{
+    auto callback = (TCPConnection stream) @safe nothrow {
+        try stream.readTimeout = timeout;
+        catch (Exception e) assert(0);
+        handle(impl, stream);
+    };
+    return listenTCP(port, callback, address);
+}
+
+/*******************************************************************************
+
+    Top-level function for handling incoming (server) requests
+
+    This is called by `listenTCP` whenever new data on the `stream` is received.
+    It simply wraps `handleThrow` as the callback needs to be `@safe nothrow`.
+
+    Params:
+      API = The type of API that will handle the call
+      api = The object that will handle the call
+      stream = The TCP stream to read data from
+
+*******************************************************************************/
+
+private void handle (API) (API api, ref TCPConnection stream) @trusted nothrow
+{
+    try
+        handleThrow(api, stream);
+    catch (Exception ex)
+    {
+        try log.error("Exception caught in handle: {}", ex);
+        // This should never happen, but if it does, at least let the user know
+        catch (Exception e)
+        {
+            printf("[%s:%d] Error while logging an error: %.*s\n",
+                   __FILE__.ptr, __LINE__, cast(int) ex.msg.length, ex.msg.ptr);
+        }
+        stream.close();
+    }
+}
+
+/*******************************************************************************
+
+    Handle incoming (server) requests
+
+    This function will read from `stream`, find which function of `api` to call,
+    deserialize the parameters accordingly, then call `api`.
+    If anything goes wrong (deserialization failed, wrong method, etc...),
+    a trace message will be logged and `api` won't be called.
+
+    Params:
+      API = The type of API that will handle the call
+      api = The object that will handle the call
+      stream = The TCP stream to read data from
+
+*******************************************************************************/
+
+private void handleThrow (API) (scope API api, ref TCPConnection stream)
+    @trusted
+{
+    static string[Hash] lookup;
+    if (lookup.length == 0)
+    {
+        static foreach (member; __traits(allMembers, API))
+            static foreach (ovrld; __traits(getOverloads, API, member))
+                lookup[hashFull(ovrld.mangleof)] = ovrld.mangleof;
+    }
+
+    log.info("[{}] Handling a new request: {}", stream.peerAddress, stream.leastSize());
+    Hash methodbin;
+    stream.read(methodbin[]);
+    const method = methodbin in lookup;
+    if (method is null)
+    {
+        log.error("[{}] Calling out of range method: {}",
+                  stream.peerAddress, methodbin);
+        return;
+    }
+
+    ubyte[1024] buffer = void;
+    scope DeserializeDg reader = (size_t size) @safe
+    {
+        assert(size < buffer.length);
+        scope (failure) assert(0);
+        stream.read(buffer[0 .. size]);
+        return buffer[0 .. size];
+    };
```

Comment:
Hum I wonder if 1024 will not cause problem. Looks like an obvious DoS vector.

Reply:
Let's just throw on that. If 1024 is enough for our arguments there wont be any problem.

Reply:
Sounds good.